### PR TITLE
Prepare for liftA2 being exported from Prelude

### PR DESCRIPTION
**Pull request by googleson78**
CLC has approved the proposal to export `liftA2` from `Prelude`
(https://github.com/haskell/core-libraries-committee/issues/50).

This is a PR to future proof for that change, while minimising warnings.

### CODE REVIEW

File: containers/src/Data/IntMap/Strict/Internal.hs
Comment (googleson78) [inline containers/src/Data/IntMap/Strict/Internal.hs:348]: Is the `liftA2` explicit import here necessary on some older ghcs?

Comment (googleson78) [reply containers/src/Data/IntMap/Strict/Internal.hs:348]: Kind of deleted it without thinking..

**Comment by Bodigrim** [comment]
@googleson78 could you please fix CI?

**Comment by googleson78** [comment]
I tried to take into account ghc 8.0.2 this time around.

**Comment by Bodigrim** [comment]
@treeowl @sjakobi could you please take a look? We need this merged before proceeding with https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8449

**Review by sjakobi — APPROVED** [review]

**Comment by Bodigrim** [comment]
@treeowl just a gentle reminder. Otherwise it blocks [the respective GHC MR](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8449).

File: containers/src/Data/IntMap/Internal.hs
Comment (andreasabel) [inline containers/src/Data/IntMap/Internal.hs:297]: Using `hiding` isn't best practice.  Please change to explicit import list.
(The exception is `import Prelude hiding (...)` because explicit import lists for `Prelude` are looong.)

On a second thought, I think that placing the `import Control.Applicative (liftA2)` under a `!MIN_VERSION_base` constraint, and not changing anything else, would be the best remedy here.

File: containers/src/Data/IntMap/Strict/Internal.hs
Comment (andreasabel) [inline containers/src/Data/IntMap/Strict/Internal.hs:348]: Ditto.

File: containers/src/Data/Sequence/Internal.hs
Comment (andreasabel) [inline containers/src/Data/Sequence/Internal.hs:202]: ~~Expand `..` to precise list of imports.~~
Maybe there is a solution not needing a `(..)` import?

File: containers/src/Data/Tree.hs
Comment (andreasabel) [inline containers/src/Data/Tree.hs:57]: Ditto.

File: containers/src/Data/Graph.hs
Comment (andreasabel) [inline containers/src/Data/Graph.hs:95]: I am surprised this line is needed.

File: containers/src/Data/Tree.hs
Comment (andreasabel) [reply containers/src/Data/Tree.hs:57]: Why is any change to this file needed at all?

**Review by andreasabel — CHANGES_REQUESTED** [review]
Import-styles `hiding` and `(..)` should be avoided where possible.  Maybe `MIN_VERSION_base` can help.

**Comment by Bodigrim** [comment]
> Import-styles `hiding` and `(..)` should be avoided where possible. Maybe `MIN_VERSION_base` can help.

I would prefer `hiding` to additional CPP any day, especially for a boot library, because `MIN_VERSION_base` is sometimes wrong in nightly GHC builds. Let's wait for @treeowl opinion.

**Comment by andreasabel** [comment]
It is somehow strange that changes are necessary at all here.  The PR description isn't very explicit about what problems it addresses.  Let me guess: GHC is complaining whenever `liftA2` is explicitly imported from somewhere other than `Prelude`, because it is already implicitly imported from `Prelude`.
That's silly by GHC.  That warning logic is simply odd.  All the energy developers are spending into fixing GHC's silly _unused import_ warnings should maybe go into fixing GHC.  See https://www.snoyman.com/blog/2020/11/haskell-bad-parts-2/ :
> ... we all turn on `-Wall`. Because we're good kids. We want to do the right thing. And this, of course, turns on unused import warnings. And because each new release of GHC and every library on Hackage likes to mess with us, we are constantly exporting new identifiers from different modules.

> The amount of time I have spent adding weird hacks to account for the fact that `<>` is suddenly exposed from `Prelude`, and therefore my `import Data.Monoid ((<>))` isn't necessary, is obscene. The introduction of fiddly CPP to work around this sucks. In fact, this all sucks. It's horrible.

In the light of this, maybe the better patch here is to add `-fno-warn-unused-imports` to the `ghc-options` of the `.cabal` file (resp. `-Wno-unused-imports` for GHC >= 8) until the oddity in GHC is fixed.

**Comment by Bodigrim** [comment]
@treeowl could you please opine on this? We are blocked for more than a month.

**Comment by treeowl** [comment]
I agree with @andreasabel that the `hiding` approach isn't great. Can we avoid it without too much CPP?

**Comment by Bodigrim** [comment]
@googleson78 any chance you can pick it up again? If you are busy, I can give it a try myself.

**Comment by googleson78** [comment]
If we want to avoid the hiding personally I'd prefer just disabling the warning, as Andreas suggested in 
> In the light of this, maybe the better patch here is to add -fno-warn-unused-imports to the ghc-options of the .cabal file (resp. -Wno-unused-imports for GHC >= 8) until the oddity in GHC is fixed.

**Comment by Bodigrim** [comment]
@treeowl how do you feel about throwing `-fno-warn-unused-imports`? I agree that it's probably the least evil.

**Comment by treeowl** [comment]
We can definitely disable unused import warnings for older GHC to unstick.

**Comment by googleson78** [comment]
I reverted the previous changes and disabled that warning globally for `containers`.

**Comment by treeowl** [comment]
Will review after CI

**Review by andreasabel — APPROVED** [review]
Great, thanks!

**Comment by treeowl** [comment]
I meant you can turn off the warnings for old versions. Why do we need to turn them off for all? Is it really too painful otherwise?

**Comment by googleson78** [comment]
It would need to be turned on for versions **newer** than whatever ghc/base version ships with the `liftA2` export change.

**Comment by treeowl** [comment]
Is there a way to confine the mess to a separate module?

**Comment by Bodigrim** [comment]
@treeowl @sjakobi could you please take ownership of this issue and merge any solution which suits you? The upstream GHC is blocked on this for two months. Asking a volunteer to explore completely orthogonal solutions just to be rejected on a whim strikes my as unreasonable.

I'd suggest to merge any approach (e. g., the current state of PR) to unblock GHC, and then you can polish it until `containers` release in whatever direction you wish.

**Comment by sjakobi** [comment]
I don't consider myself a maintainer of `containers` at this point, so I take no responsibility for this PR.

Based on the discussion here, I'd suggest to add `{-# OPTIONS_GHC -fno-warn-unused-imports #-}` and explanatory comments in each of the affected modules.

**Comment by treeowl** [comment]
Ugh. I'll do something tonight or tomorrow.

**Comment by Bodigrim** [comment]
@treeowl ping.

**Comment by treeowl** [comment]
> @treeowl ping.

Time scheduled for this today.

**Comment by treeowl** [comment]
I'm attempting an alternative approach in #852, which plays "Let's pretend `liftA2` has always been in the `Prelude`.`

**Comment by treeowl** [comment]
I hate the approach in #852 a bit less, so I'm doing that instead.

**Comment by Bodigrim** [comment]
Thanks!

**Pull request closed without merging**
